pxr/imaging/plugin/hdRpr/rprApi.cpp

```diff
@@ -1070,40 +1087,123 @@ class HdRprApiImpl {
+        auto& wvm = m_hdCamera->GetViewMatrix();
+        auto aspectRatio = double(m_viewportSize[0]) / m_viewportSize[1];
 
         GfVec3f eye(iwvm[3][0], iwvm[3][1], iwvm[3][2]);
         GfVec3f up(wvm[0][1], wvm[1][1], wvm[2][1]);
         GfVec3f n(wvm[0][2], wvm[1][2], wvm[2][2]);
         GfVec3f at(eye - n);
         RPR_ERROR_CHECK(rprCameraLookAt(camera, eye[0], eye[1], eye[2], at[0], at[1], at[2], up[0], up[1], up[2]), "Fail to set camera Look At");
 
-        bool isOrthographic = round(m_cameraProjectionMatrix[3][3]) == 1.0;
-        if (isOrthographic) {
-            GfVec3f ndcTopLeft(-1.0f, 1.0f, 0.0f);
-            GfVec3f nearPlaneTrace = m_cameraProjectionMatrix.GetInverse().Transform(ndcTopLeft);
+        GfRange1f clippingRange(0.01f, 100000000.0f);
+        m_hdCamera->GetClippingRange(&clippingRange);
+        RPR_ERROR_CHECK(rprCameraSetNearPlane(camera, clippingRange.GetMin()), "Failed to set camera near plane");
+        RPR_ERROR_CHECK(rprCameraSetFarPlane(camera, clippingRange.GetMax()), "Failed to set camera far plane");
+
+        double shutterOpen = 0.0;
+        double shutterClose = 0.0;
+        if (!instantaneousShutter.value) {
+            m_hdCamera->GetShutterOpen(&shutterOpen);
+            m_hdCamera->GetShutterClose(&shutterClose);
+        }
+        // XXX (RPR): shutter close/open is a time value, but RPR accepts some abstract coefficient in [0, 1] range
+        RPR_ERROR_CHECK(rprCameraSetExposure(camera, std::max(shutterClose - shutterOpen, 0.0)), "Failed to set camera exposure");
+
+        float sensorWidth;
+        float sensorHeight;
+        float focalLength;
+
+        GfVec2f apertureSize;
+        TfToken projectionType;
+        if (m_hdCamera->GetFocalLength(&focalLength) &&
+            m_hdCamera->GetApertureSize(&apertureSize) &&
+            m_hdCamera->GetProjectionType(&projectionType)) {
+            ApplyAspectRatioPolicy(m_viewportSize, aspectRatioPolicy.value, apertureSize);
+            sensorWidth = apertureSize[0];
+            sensorHeight = apertureSize[1];
+
+            float zNear = clippingRange.GetMin();
+            float zFar = clippingRange.GetMax();
+            if (projectionType == UsdGeomTokens->orthographic) {
+                m_cameraProjectionMatrix.SetIdentity();
+                m_cameraProjectionMatrix[0][0] = 2.0f / apertureSize[0];
+                m_cameraProjectionMatrix[1][1] = 2.0f / apertureSize[1];
+                m_cameraProjectionMatrix[2][2] = -2.0f / (zFar - zNear);
+                m_cameraProjectionMatrix[3][2] = -(zFar + zNear) / (zFar - zNear);
+            } else {
+                if (projectionType != UsdGeomTokens->perspective) {
+                    TF_CODING_ERROR("Unexpected projection type: %s. Fallback value: perspective", projectionType.GetText());
+                    projectionType = UsdGeomTokens->perspective;
+                }
+
+                float range = (apertureSize[1] * 0.5f) / focalLength * zNear;
```

Comment:
Do you think these calculations should be in rprCamera.cpp?

Reply:
Yeah, it should be there

Reply:
I will refactor it later today

Reply:
These calculations could not be moved to HdRprCamera because they rely on the viewport size which currently available only in HdRprApi and HdRprRenderPass. Right now, HdRprCamera is a simple sync&store entity

Reply:
Ah.  That makes sense.